File: frontend/tests/contract/api/runs.contract.test.ts
// Runs API Contract Tests
// Validates API integration against contracts/openapi.yaml

import { describe, it, expect, beforeAll, afterAll, afterEach } from 'vitest';
import { setupServer } from 'msw/node';
import { http, HttpResponse } from 'msw';

// API client (not yet implemented - tests will fail)
import { apiClient } from '@/services/api';
import type {
  RunDetailsResponse,
  BatchItemsResponse,
  ErrorResponse
} from '@/types/api';

// =============================================================================
// MSW Mock Server Setup
// =============================================================================

const server = setupServer();

beforeAll(() => server.listen({ onUnhandledRequest: 'error' }));
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// =============================================================================
// GET /api/runs/{runNo} - Run Details with Auto-Population
// =============================================================================

describe('GET /api/runs/{runNo}', () => {
  it('returns RunDetailsResponse schema with correct auto-populated fields', async () => {
    // Arrange: Mock successful run retrieval
    const runNo = 6000037;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}`, () => {
        return HttpResponse.json<RunDetailsResponse>({
          runNo: 6000037,
          fgItemKey: 'TSM2285A',
          fgDescription: 'Marinade, Savory',
          batches: [1, 2],
          productionDate: '2025-10-06',
          status: 'NEW',
          noOfBatches: 2
        });
      })
    );

    // Act: Fetch run details
    const response = await apiClient.getRunDetails(runNo);

    // Assert: Validate RunDetailsResponse schema
    expect(response).toBeDefined();
    expect(response.runNo).toBe(6000037);
    expect(response.fgItemKey).toBe('TSM2285A');
    expect(response.fgDescription).toBeTruthy();
    expect(response.batches).toHaveLength(2);
    expect(response.batches).toEqual([1, 2]);
    expect(response.productionDate).toMatch(/^\d{4}-\d{2}-\d{2}$/);
    expect(response.status).toBe('NEW');
    expect(response.noOfBatches).toBe(2);
  });

  it('returns 404 ErrorResponse when run not found', async () => {
    // Arrange: Mock not found response
    const runNo = 999999;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}`, () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'NOT_FOUND_RUN',
              message: `Run ${runNo} not found`,
              correlationId: 'test-correlation-id',
              details: { runNo }
            }
          },
          { status: 404 }
        );
      })
    );

    // Act & Assert: Expect 404 error
    await expect(
      apiClient.getRunDetails(runNo)
    ).rejects.toMatchObject({
      status: 404,
      error: {
        code: 'NOT_FOUND_RUN',
        message: expect.stringContaining('not found')
      }
    });
  });

  it('auto-populates FG item key from FormulaId', async () => {
    // Arrange
    const runNo = 213972;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}`, () => {
        return HttpResponse.json<RunDetailsResponse>({
          runNo: 213972,
          fgItemKey: 'TSM2285A', // From FormulaId in database
          fgDescription: 'Marinade, Savory',
          batches: [1],
          productionDate: '2025-10-06',
          status: 'NEW',
          noOfBatches: 1
        });
      })
    );

    // Act
    const response = await apiClient.getRunDetails(runNo);

    // Assert: Verify FormulaId → fgItemKey mapping
    expect(response.fgItemKey).toBeTruthy();
    expect(response.fgItemKey).toMatch(/^[A-Z0-9]+$/); // Item key format
  });

  it('returns all batch numbers for multi-batch run', async () => {
    // Arrange
    const runNo = 213935;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}`, () => {
        return HttpResponse.json<RunDetailsResponse>({
          runNo: 213935,
          fgItemKey: 'TSM2285A',
          fgDescription: 'Marinade, Savory',
          batches: [1, 2, 3], // Multiple batches
          productionDate: '2025-10-06',
          status: 'NEW',
          noOfBatches: 3
        });
      })
    );

    // Act
    const response = await apiClient.getRunDetails(runNo);

    // Assert: Verify batches array matches noOfBatches
    expect(response.batches).toHaveLength(3);
    expect(response.noOfBatches).toBe(3);
    expect(response.batches).toEqual([1, 2, 3]);
  });
});

// =============================================================================
// GET /api/runs/{runNo}/batches/{rowNum}/items - Batch Items with Weight Ranges
// =============================================================================

describe('GET /api/runs/{runNo}/batches/{rowNum}/items', () => {
  it('returns BatchItemsResponse schema with calculated weight ranges', async () => {
    // Arrange: Mock batch items retrieval
    const runNo = 6000037;
    const rowNum = 1;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: [
            {
              itemKey: 'INRICF05',
              description: 'Rice Flour (RF-0010)',
              totalNeeded: 14.24,
              pickedQty: 0,
              remainingQty: 14.24,
              weightRangeLow: 14.215,
              weightRangeHigh: 14.265,
              toleranceKG: 0.025,
              allergen: '',
              status: null
            },
            {
              itemKey: 'INSALT02',
              description: 'Salt Medium without anticaking',
              totalNeeded: 20.0,
              pickedQty: 20.025,
              remainingQty: 0,
              weightRangeLow: 19.975,
              weightRangeHigh: 20.025,
              toleranceKG: 0.025,
              allergen: '',
              status: 'Allocated'
            }
          ]
        });
      })
    );

    // Act: Fetch batch items
    const response = await apiClient.getBatchItems(runNo, rowNum);

    // Assert: Validate BatchItemsResponse schema
    expect(response).toBeDefined();
    expect(response.items).toHaveLength(2);

    const firstItem = response.items[0];
    expect(firstItem.itemKey).toBe('INRICF05');
    expect(firstItem.totalNeeded).toBe(14.24);
    expect(firstItem.remainingQty).toBe(14.24);
    expect(firstItem.weightRangeLow).toBe(14.215);
    expect(firstItem.weightRangeHigh).toBe(14.265);
    expect(firstItem.toleranceKG).toBe(0.025);

    const secondItem = response.items[1];
    expect(secondItem.status).toBe('Allocated');
    expect(secondItem.pickedQty).toBe(20.025);
  });

  it('calculates weight range low = totalNeeded - toleranceKG', async () => {
    // Arrange
    const runNo = 213972;
    const rowNum = 1;
    const totalNeeded = 20.0;
    const tolerance = 0.025;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: [
            {
              itemKey: 'INSALT02',
              description: 'Salt Medium',
              totalNeeded: totalNeeded,
              pickedQty: 0,
              remainingQty: totalNeeded,
              weightRangeLow: totalNeeded - tolerance, // 19.975
              weightRangeHigh: totalNeeded + tolerance, // 20.025
              toleranceKG: tolerance,
              allergen: '',
              status: null
            }
          ]
        });
      })
    );

    // Act
    const response = await apiClient.getBatchItems(runNo, rowNum);

    // Assert: Verify weight range calculation
    const item = response.items[0];
    expect(item.weightRangeLow).toBe(totalNeeded - tolerance);
    expect(item.weightRangeLow).toBe(19.975);
  });

  it('calculates weight range high = totalNeeded + toleranceKG', async () => {
    // Arrange
    const runNo = 213972;
    const rowNum = 1;
    const totalNeeded = 20.0;
    const tolerance = 0.025;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: [
            {
              itemKey: 'INSALT02',
              description: 'Salt Medium',
              totalNeeded: totalNeeded,
              pickedQty: 0,
              remainingQty: totalNeeded,
              weightRangeLow: totalNeeded - tolerance,
              weightRangeHigh: totalNeeded + tolerance, // 20.025
              toleranceKG: tolerance,
              allergen: '',
              status: null
            }
          ]
        });
      })
    );

    // Act
    const response = await apiClient.getBatchItems(runNo, rowNum);

    // Assert: Verify weight range calculation
    const item = response.items[0];
    expect(item.weightRangeHigh).toBe(totalNeeded + tolerance);
    expect(item.weightRangeHigh).toBe(20.025);
  });

  it('calculates remaining quantity = totalNeeded - pickedQty', async () => {
    // Arrange
    const runNo = 213972;
    const rowNum = 1;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: [
            {
              itemKey: 'INSALT02',
              description: 'Salt Medium',
              totalNeeded: 20.0,
              pickedQty: 15.5,
              remainingQty: 4.5, // 20.0 - 15.5
              weightRangeLow: 19.975,
              weightRangeHigh: 20.025,
              toleranceKG: 0.025,
              allergen: '',
              status: null
            }
          ]
        });
      })
    );

    // Act
    const response = await apiClient.getBatchItems(runNo, rowNum);

    // Assert: Verify remaining quantity calculation
    const item = response.items[0];
    expect(item.remainingQty).toBe(item.totalNeeded - item.pickedQty);
    expect(item.remainingQty).toBe(4.5);
  });

  it('returns empty items array for batch with no items', async () => {
    // Arrange
    const runNo = 213972;
    const rowNum = 999;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: []
        });
      })
    );

    // Act
    const response = await apiClient.getBatchItems(runNo, rowNum);

    // Assert: Empty items array
    expect(response.items).toHaveLength(0);
  });

  it('returns 404 when batch not found', async () => {
    // Arrange
    const runNo = 213972;
    const rowNum = 999;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/${rowNum}/items`, () => {
        return HttpResponse.json<ErrorResponse>(
          {
            error: {
              code: 'NOT_FOUND_BATCH',
              message: `Batch ${rowNum} not found for run ${runNo}`,
              correlationId: 'test-correlation-id',
              details: { runNo, rowNum }
            }
          },
          { status: 404 }
        );
      })
    );

    // Act & Assert
    await expect(
      apiClient.getBatchItems(runNo, rowNum)
    ).rejects.toMatchObject({
      status: 404,
      error: {
        code: 'NOT_FOUND_BATCH'
      }
    });
  });
});

// =============================================================================
// Integration Tests - Run + Batch Items Workflow
// =============================================================================

describe('Run Details + Batch Items Workflow', () => {
  it('fetches run details then batch items in sequence', async () => {
    // Arrange: Mock both endpoints
    const runNo = 213972;

    server.use(
      http.get(`http://localhost:7075/api/runs/${runNo}`, () => {
        return HttpResponse.json<RunDetailsResponse>({
          runNo: 213972,
          fgItemKey: 'TSM2285A',
          fgDescription: 'Marinade, Savory',
          batches: [1, 2],
          productionDate: '2025-10-06',
          status: 'NEW',
          noOfBatches: 2
        });
      }),
      http.get(`http://localhost:7075/api/runs/${runNo}/batches/1/items`, () => {
        return HttpResponse.json<BatchItemsResponse>({
          items: [
            {
              itemKey: 'INSALT02',
              description: 'Salt Medium',
              totalNeeded: 20.0,
              pickedQty: 0,
              remainingQty: 20.0,
              weightRangeLow: 19.975,
              weightRangeHigh: 20.025,
              toleranceKG: 0.025,
              allergen: '',
              status: null
            }
          ]
        });
      })
    );

    // Act: Execute workflow
    const runDetails = await apiClient.getRunDetails(runNo);
    const batchItems = await apiClient.getBatchItems(runNo, runDetails.batches[0]);

    // Assert: Workflow completed successfully
    expect(runDetails.runNo).toBe(213972);
    expect(runDetails.batches).toContain(1);
    expect(batchItems.items).toHaveLength(1);
    expect(batchItems.items[0].itemKey).toBe('INSALT02');
  });
});

File: backend/tests/contract/runs_contract_test.rs
// Production Runs Contract Tests
// Tests validate implementation against contracts/openapi.yaml

use axum::http::StatusCode;
use serde_json::json;

mod test_helpers;
use test_helpers::*;

// =============================================================================
// GET /api/runs/{runNo} - Run Details with Auto-Population
// =============================================================================

#[tokio::test]
async fn test_get_run_details_success() {
    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act
    let response = app
        .get("/api/runs/6000037")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert: Response matches RunDetailsResponse schema
    assert_eq!(response.status(), StatusCode::OK);

    let body: RunDetailsResponse = response.json().await;

    // Validate auto-populated fields from database
    assert_eq!(body.run_no, 6000037);
    assert_eq!(body.fg_item_key, "TSM2285A"); // From FormulaId
    assert_eq!(body.fg_description, "Marinade, Savory"); // From FormulaDesc
    assert_eq!(body.batches, vec![1, 2]); // All RowNum values
    assert_eq!(body.production_date, "2025-10-06"); // From RecDate
    assert_eq!(body.status, "NEW");
    assert_eq!(body.no_of_batches, 2);
}

#[tokio::test]
async fn test_get_run_details_not_found() {
    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act
    let response = app
        .get("/api/runs/9999999")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert: 404 with ErrorResponse schema
    assert_eq!(response.status(), StatusCode::NOT_FOUND);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "DB_RECORD_NOT_FOUND");
    assert!(body.error.message.contains("9999999"));
    assert!(!body.error.correlation_id.is_empty());
}

#[tokio::test]
async fn test_get_run_details_unauthorized() {
    // Arrange
    let app = create_test_app().await;

    // Act: Request without JWT token
    let response = app.get("/api/runs/6000037").send().await;

    // Assert
    assert_eq!(response.status(), StatusCode::UNAUTHORIZED);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "AUTH_INVALID_TOKEN");
}

// =============================================================================
// GET /api/runs/{runNo}/batches/{rowNum}/items - Batch Items with Weight Range
// =============================================================================

#[tokio::test]
async fn test_get_batch_items_success() {
    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act
    let response = app
        .get("/api/runs/6000037/batches/1/items")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert: Response matches BatchItemsResponse schema
    assert_eq!(response.status(), StatusCode::OK);

    let body: BatchItemsResponse = response.json().await;
    assert!(!body.items.is_empty());

    // Validate first item
    let item = &body.items[0];
    assert_eq!(item.item_key, "INRICF05");
    assert_eq!(item.description, "Rice Flour (RF-0010)");
    assert_eq!(item.total_needed, 14.24);
    assert_eq!(item.picked_qty, 0.0);
    assert_eq!(item.remaining_qty, 14.24);

    // Validate weight range calculation
    // weightRangeLow = ToPickedPartialQty - INMAST.User9
    // weightRangeHigh = ToPickedPartialQty + INMAST.User9
    assert_eq!(item.weight_range_low, 14.215); // 14.24 - 0.025
    assert_eq!(item.weight_range_high, 14.265); // 14.24 + 0.025
    assert_eq!(item.tolerance_kg, 0.025);
}

#[tokio::test]
async fn test_get_batch_items_with_picked_items() {
    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act
    let response = app
        .get("/api/runs/6000037/batches/1/items")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::OK);

    let body: BatchItemsResponse = response.json().await;

    // Find picked item (if any exist in test data)
    let picked_item = body.items.iter().find(|item| item.picked_qty > 0.0);

    if let Some(item) = picked_item {
        // Validate status for picked item
        assert_eq!(item.status, Some("Allocated".to_string()));
        assert!(item.remaining_qty < item.total_needed);

        // Verify picked_qty is within tolerance
        let tolerance_check = (item.picked_qty - item.total_needed).abs() <= item.tolerance_kg;
        assert!(
            tolerance_check,
            "Picked quantity should be within tolerance"
        );
    }
}

#[tokio::test]
async fn test_get_batch_items_not_found() {
    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act: Invalid batch number
    let response = app
        .get("/api/runs/6000037/batches/999/items")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert
    assert_eq!(response.status(), StatusCode::NOT_FOUND);

    let body: ErrorResponse = response.json().await;
    assert_eq!(body.error.code, "DB_RECORD_NOT_FOUND");
}

#[tokio::test]
async fn test_get_batch_items_composite_key_validation() {
    // This test validates that the implementation uses composite keys
    // (RunNo, RowNum) correctly, not artificial IDs

    // Arrange
    let app = create_test_app().await;
    let token = get_test_auth_token(&app).await;

    // Act: Get items for different batches of same run
    let response_batch1 = app
        .get("/api/runs/6000037/batches/1/items")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    let response_batch2 = app
        .get("/api/runs/6000037/batches/2/items")
        .header("Authorization", format!("Bearer {}", token))
        .send()
        .await;

    // Assert: Both batches should exist but have different items
    assert_eq!(response_batch1.status(), StatusCode::OK);
    assert_eq!(response_batch2.status(), StatusCode::OK);

    let body1: BatchItemsResponse = response_batch1.json().await;
    let body2: BatchItemsResponse = response_batch2.json().await;

    // Validate composite key usage (items are different per batch)
    // Constitutional requirement: No artificial IDs, use (RunNo, RowNum, LineId)
    assert!(body1.items.len() > 0);
    assert!(body2.items.len() > 0);
}

// =============================================================================
// Type Definitions (matching OpenAPI schemas)
// =============================================================================

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
struct RunDetailsResponse {
    #[serde(rename = "runNo")]
    run_no: i32,
    #[serde(rename = "fgItemKey")]
    fg_item_key: String,
    #[serde(rename = "fgDescription")]
    fg_description: String,
    batches: Vec<i32>,
    #[serde(rename = "productionDate")]
    production_date: String,
    status: String,
    #[serde(rename = "noOfBatches")]
    no_of_batches: i32,
}

#[derive(Debug, Serialize, Deserialize)]
struct BatchItemsResponse {
    items: Vec<BatchItemDTO>,
}

#[derive(Debug, Serialize, Deserialize)]
struct BatchItemDTO {
    #[serde(rename = "itemKey")]
    item_key: String,
    description: String,
    #[serde(rename = "totalNeeded")]
    total_needed: f64,
    #[serde(rename = "pickedQty")]
    picked_qty: f64,
    #[serde(rename = "remainingQty")]
    remaining_qty: f64,
    #[serde(rename = "weightRangeLow")]
    weight_range_low: f64,
    #[serde(rename = "weightRangeHigh")]
    weight_range_high: f64,
    #[serde(rename = "toleranceKG")]
    tolerance_kg: f64,
    allergen: String,
    status: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct ErrorResponse {
    error: ErrorDetails,
}

#[derive(Debug, Serialize, Deserialize)]
struct ErrorDetails {
    code: String,
    message: String,
    #[serde(rename = "correlationId")]
    correlation_id: String,
    details: Option<serde_json::Value>,
}
